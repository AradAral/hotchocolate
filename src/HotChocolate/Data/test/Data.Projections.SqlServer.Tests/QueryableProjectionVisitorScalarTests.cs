using System.Threading.Tasks;
using HotChocolate.Execution;
using Xunit;

namespace HotChocolate.Data.Projections
{
    public class QueryableProjectionVisitorScalarTests
    {
        private static readonly Foo[] _fooEntities =
        {
            new Foo { Bar = true, Baz = "a" }, new Foo { Bar = false, Baz = "b" }
        };

        private static readonly FooNullable[] _fooNullableEntities =
        {
            new FooNullable { Bar = true, Baz = "a" },
            new FooNullable { Bar = null, Baz = null },
            new FooNullable { Bar = false, Baz = "c" }
        };

        private readonly SchemaCache _cache = new SchemaCache();

        [Fact]
        public async Task Create_ProjectsTwoProperties_Expression()
        {
            // arrange
            IRequestExecutor tester = _cache.CreateSchema(_fooEntities);

            // act
            // assert
            IExecutionResult res1 = await tester.ExecuteAsync(
                QueryRequestBuilder.New()
                    .SetQuery("{ root{ bar baz }}")
                    .Create());

            res1.MatchSqlSnapshot();
        }

        [Fact]
        public async Task Create_ProjectsOneProperty_Expression()
        {
            // arrange
            IRequestExecutor tester = _cache.CreateSchema(_fooEntities);

            // act
            // assert
            IExecutionResult res1 = await tester.ExecuteAsync(
                QueryRequestBuilder.New()
                    .SetQuery("{ root{ bar baz }}")
                    .Create());

            res1.MatchSqlSnapshot();
        }

        public class Foo
        {
            public int Id { get; set; }

            public bool Bar { get; set; }

            public string Baz { get; set; }
        }

        public class FooNullable
        {
            public int Id { get; set; }

            public bool? Bar { get; set; }

            public string? Baz { get; set; }
        }
    }
}
