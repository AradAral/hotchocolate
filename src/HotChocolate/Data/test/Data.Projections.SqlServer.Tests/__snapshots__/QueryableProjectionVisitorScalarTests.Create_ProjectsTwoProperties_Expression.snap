{
  "errors": [
    {
      "message": "Unexpected Execution Error",
      "locations": [
        {
          "line": 1,
          "column": 3
        }
      ],
      "path": [
        "root"
      ],
      "extensions": {
        "message": "Object reference not set to an instance of an object.",
        "stackTrace": "   at HotChocolate.Data.Projections.SelectionVisitor\u00601.Visit(IOutputField field, TContext context) in /home/sepa/kot/hotchocolate/src/HotChocolate/Data/src/Data/Projections/SelectionVisitor.cs:line 50\n   at HotChocolate.Data.Projections.ProjectionVisitor\u00601.Visit(TContext context) in /home/sepa/kot/hotchocolate/src/HotChocolate/Data/src/Data/Projections/ProjectionVisitor.cs:line 22\n   at HotChocolate.Data.Projections.Expressions.QueryableProjectionConvention.\u003CCreateExecutor\u003Eg__ExecuteAsync|2_1[TEntityType](FieldDelegate next, IMiddlewareContext context) in /home/sepa/kot/hotchocolate/src/HotChocolate/Data/src/Data/Projections/Expressions/QueryableProjectionConvention.cs:line 50\n   at HotChocolate.Data.Projections.ProjectionVisitorTestBase.\u003C\u003Ec__DisplayClass6_1\u00601.\u003C\u003CCreateSchema\u003Eb__5\u003Ed.MoveNext() in /home/sepa/kot/hotchocolate/src/HotChocolate/Data/test/Data.Projections.SqlServer.Tests/ProjectionVisitorTestBase.cs:line 65\n--- End of stack trace from previous location ---\n   at HotChocolate.Execution.Processing.ResolverTask.ExecuteResolverPipelineAsync() in /home/sepa/kot/hotchocolate/src/HotChocolate/Core/src/Execution/Processing/ResolverTask.cs:line 65\n   at HotChocolate.Execution.Processing.ResolverTask.TryExecuteAsync() in /home/sepa/kot/hotchocolate/src/HotChocolate/Core/src/Execution/Processing/ResolverTask.cs:line 51"
      }
    }
  ],
  "data": {
    "root": null
  }
}
